substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/service/ServiceSupport.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2024, 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2024, 2024, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core.genscavenge.service;
+
+import com.oracle.svm.core.heap.GCCause;
+import com.oracle.svm.core.Uninterruptible;
+import jdk.graal.compiler.api.replacements.Fold;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+import org.graalvm.word.UnsignedWord;
+
+public class ServiceSupport {
+    private GcNotifier gcNotifier;
+    private ServiceThread serviceThread;
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    public ServiceSupport() {
+        gcNotifier = new GcNotifier();
```

Comment:
`GcNotifier` should be a conditionally registered image singleton. If the service thread infrastructure is used for other use cases in the future, there is guarantee that a `GcNotifier` is actually present.

Reply:
Hmm I'm not sure I fully understand.  Do you mean that it should be conditionally registered here inside `ServiceSupport`?  Why would GcNotifier not be present if the `ServiceFeature` has been included?

Reply:
The service/notification thread may be used by multiple features (this is not the case at the moment but it is very likely that this will be the case in the future). Not all features are necessarily present in a Native Image, so we should implement generic infrastructure (like the service/notification thread) in a way that it still works even if some features are not available.

If the GC notifier is an image singleton, this can be achieve easily by calling `ImageSingletons.contains(...)` in the notification thread to check if the GC notifier is present. If so, the thread can do the corresponding processing work.

Reply:
Ok I've made `GcNotifier` an image singleton that is registered if `GcNotificationFeature` is included at build time with `--enable-monitoring=gcnotif`. 

`NotificationThreadFeature` is included depending on whether `GcNotificationFeature` or other future `*NotificationFeature`s exist. 

`NotificationThread` will check, in it's loop, whether each type of notification feature exists before trying to send any notifications of that type.